File: crates/piggy-pigpen/src/crypto.rs
//! pigpen-v1 crypto suite (RFC 0008 §4), pure-Rust / wasm-buildable.

use chacha20poly1305::aead::Aead;
use chacha20poly1305::{ChaCha20Poly1305, Key, KeyInit, Nonce};
use hkdf::Hkdf;
use hmac::{Hmac, Mac};
use p256::elliptic_curve::sec1::ToEncodedPoint;
use rand_core::{OsRng, RngCore};
use sha2::Sha256;

use crate::{Error, Result};

type HmacSha256 = Hmac<Sha256>;

pub const FILE_KEY_LEN: usize = 16; // RFC 0008 §4.1
pub const PAYLOAD_NONCE_LEN: usize = 16; // RFC 0008 §4.5
pub const STREAM_CHUNK: usize = 64 * 1024;
pub const TAG_LEN: usize = 16;

const INFO_PAYLOAD: &[u8] = b"pigpen-v1 payload";
const INFO_HEADER: &[u8] = b"pigpen-v1 header";
const INFO_P256: &[u8] = b"pigpen-v1 piv-p256";
const INFO_X25519: &[u8] = b"pigpen-v1 x25519";

fn hkdf32(ikm: &[u8], salt: &[u8], info: &[u8]) -> [u8; 32] {
    let hk = Hkdf::<Sha256>::new(Some(salt), ikm);
    let mut okm = [0u8; 32];
    hk.expand(info, &mut okm).expect("32 is a valid HKDF length");
    okm
}

fn aead_seal(key: &[u8; 32], nonce: &[u8; 12], plaintext: &[u8]) -> Result<Vec<u8>> {
    let aead = ChaCha20Poly1305::new(Key::from_slice(key));
    aead.encrypt(Nonce::from_slice(nonce), plaintext)
        .map_err(|e| Error::Crypto(format!("aead seal: {e}")))
}

fn aead_open(key: &[u8; 32], nonce: &[u8; 12], ciphertext: &[u8]) -> Result<Vec<u8>> {
    let aead = ChaCha20Poly1305::new(Key::from_slice(key));
    aead.decrypt(Nonce::from_slice(nonce), ciphertext)
        .map_err(|e| Error::Crypto(format!("aead open: {e}")))
}

const ZERO_NONCE: [u8; 12] = [0u8; 12];

pub fn random_file_key() -> [u8; FILE_KEY_LEN] {
    let mut fk = [0u8; FILE_KEY_LEN];
    OsRng.fill_bytes(&mut fk);
    fk
}

// --- X25519 wrap (RFC 0008 §4.4) ----------------------------------------

pub fn wrap_x25519(file_key: &[u8], recipient_pub: &[u8]) -> Result<Vec<u8>> {
    let rpub: [u8; 32] = recipient_pub
        .try_into()
        .map_err(|_| Error::Crypto("x25519 recipient pubkey must be 32 bytes".into()))?;
    let recipient = x25519_dalek::PublicKey::from(rpub);

    let esk = x25519_dalek::EphemeralSecret::random_from_rng(OsRng);
    let epk = x25519_dalek::PublicKey::from(&esk);
    let shared = esk.diffie_hellman(&recipient);

    let mut salt = Vec::with_capacity(64);
    salt.extend_from_slice(epk.as_bytes());
    salt.extend_from_slice(&rpub);
    let kw = hkdf32(shared.as_bytes(), &salt, INFO_X25519);
    let ct = aead_seal(&kw, &ZERO_NONCE, file_key)?;

    let mut blob = Vec::with_capacity(32 + ct.len());
    blob.extend_from_slice(epk.as_bytes());
    blob.extend_from_slice(&ct);
    Ok(blob)
}

pub fn unwrap_x25519(blob: &[u8], recipient_pub: &[u8], recipient_sec: &[u8]) -> Result<Vec<u8>> {
    if blob.len() != 32 + FILE_KEY_LEN + TAG_LEN {
        return Err(Error::Crypto("bad x25519 wrap length".into()));
    }
    let (epk_b, ct) = blob.split_at(32);
    let epk: [u8; 32] = epk_b.try_into().unwrap();
    let sec: [u8; 32] = recipient_sec
        .try_into()
        .map_err(|_| Error::Crypto("x25519 secret must be 32 bytes".into()))?;

    let sk = x25519_dalek::StaticSecret::from(sec);
    let shared = sk.diffie_hellman(&x25519_dalek::PublicKey::from(epk));

    let mut salt = Vec::with_capacity(64);
    salt.extend_from_slice(&epk);
    salt.extend_from_slice(recipient_pub);
    let kw = hkdf32(shared.as_bytes(), &salt, INFO_X25519);
    aead_open(&kw, &ZERO_NONCE, ct)
}

// --- P-256 wrap (RFC 0008 §4.3) -----------------------------------------

pub fn wrap_p256(file_key: &[u8], recipient_compressed: &[u8]) -> Result<Vec<u8>> {
    let recipient = p256::PublicKey::from_sec1_bytes(recipient_compressed)
        .map_err(|e| Error::Crypto(format!("bad P-256 recipient: {e}")))?;

    let esk = p256::ecdh::EphemeralSecret::random(&mut OsRng);
    let epk = esk.public_key();
    let epk_compressed = epk.to_encoded_point(true);
    let shared = esk.diffie_hellman(&recipient);

    let mut salt = Vec::with_capacity(66);
    salt.extend_from_slice(epk_compressed.as_bytes());
    salt.extend_from_slice(recipient_compressed);
    let kw = hkdf32(shared.raw_secret_bytes(), &salt, INFO_P256);
    let ct = aead_seal(&kw, &ZERO_NONCE, file_key)?;

    let mut blob = Vec::with_capacity(33 + ct.len());
    blob.extend_from_slice(epk_compressed.as_bytes());
    blob.extend_from_slice(&ct);
    Ok(blob)
}

/// Unwrap a P-256 stanza given the 32-byte ECDH X-coordinate the oracle
/// (the card) returns for `(slot-9D · epk)`.
pub fn unwrap_p256_with_shared(
    blob: &[u8],
    recipient_compressed: &[u8],
    shared_x: &[u8],
) -> Result<Vec<u8>> {
    if blob.len() != 33 + FILE_KEY_LEN + TAG_LEN {
        return Err(Error::Crypto("bad p256 wrap length".into()));
    }
    let (epk_b, ct) = blob.split_at(33);
    let mut salt = Vec::with_capacity(66);
    salt.extend_from_slice(epk_b);
    salt.extend_from_slice(recipient_compressed);
    let kw = hkdf32(shared_x, &salt, INFO_P256);
    aead_open(&kw, &ZERO_NONCE, ct)
}

/// The compressed ephemeral pubkey carried in a P-256 wrap blob — the
/// `partner_epk` an [`crate::EcdhOracle`] needs.
pub fn p256_wrap_epk(blob: &[u8]) -> Result<&[u8]> {
    if blob.len() < 33 {
        return Err(Error::Crypto("p256 wrap too short".into()));
    }
    Ok(&blob[..33])
}

// --- Payload STREAM (RFC 0008 §4.5) -------------------------------------

pub fn seal_payload(file_key: &[u8], plaintext: &[u8]) -> Result<Vec<u8>> {
    let mut nonce = [0u8; PAYLOAD_NONCE_LEN];
    OsRng.fill_bytes(&mut nonce);
    let stream_key = hkdf32(file_key, &nonce, INFO_PAYLOAD);
    let aead = ChaCha20Poly1305::new(Key::from_slice(&stream_key));

    let mut out = nonce.to_vec();
    let mut i = 0usize;
    loop {
        let start = i * STREAM_CHUNK;
        let end = (start + STREAM_CHUNK).min(plaintext.len());
        let last = end >= plaintext.len();
        let chunk = &plaintext[start..end];
        let ct = aead
            .encrypt(Nonce::from_slice(&stream_nonce(i as u64, last)), chunk)
            .map_err(|e| Error::Crypto(format!("payload seal: {e}")))?;
        out.extend_from_slice(&ct);
        if last {
            break;
        }
        i += 1;
    }
    Ok(out)
}

pub fn open_payload(file_key: &[u8], payload: &[u8]) -> Result<Vec<u8>> {
    if payload.len() < PAYLOAD_NONCE_LEN {
        return Err(Error::Crypto("payload shorter than nonce".into()));
    }
    let (nonce, body) = payload.split_at(PAYLOAD_NONCE_LEN);
    let stream_key = hkdf32(file_key, nonce, INFO_PAYLOAD);
    let aead = ChaCha20Poly1305::new(Key::from_slice(&stream_key));

    let enc_chunk = STREAM_CHUNK + TAG_LEN;
    let mut out = Vec::new();
    let mut i = 0usize;
    loop {
        let start = i * enc_chunk;
        if start >= body.len() {
            return Err(Error::Crypto("truncated payload (no final chunk)".into()));
        }
        let end = (start + enc_chunk).min(body.len());
        let last = end >= body.len();
        let chunk = &body[start..end];
        let plain = aead
            .decrypt(Nonce::from_slice(&stream_nonce(i as u64, last)), chunk)
            .map_err(|e| Error::Crypto(format!("payload chunk {i}: {e}")))?;
        out.extend_from_slice(&plain);
        if last {
            break;
        }
        i += 1;
    }
    Ok(out)
}

/// 12-byte age STREAM nonce: 11-byte big-endian counter + 1-byte last flag.
fn stream_nonce(counter: u64, last: bool) -> [u8; 12] {
    let mut n = [0u8; 12];
    n[3..11].copy_from_slice(&counter.to_be_bytes());
    if last {
        n[11] = 0x01;
    }
    n
}

// --- Header MAC (RFC 0008 §4.6) -----------------------------------------

pub fn header_mac(file_key: &[u8], canonical_header: &[u8]) -> [u8; 32] {
    let km = hkdf32(file_key, &[], INFO_HEADER);
    let mut mac = <HmacSha256 as Mac>::new_from_slice(&km).expect("hmac key");
    mac.update(canonical_header);
    mac.finalize().into_bytes().into()
}

pub fn verify_mac(file_key: &[u8], canonical_header: &[u8], expected: &[u8]) -> bool {
    let km = hkdf32(file_key, &[], INFO_HEADER);
    let mut mac = <HmacSha256 as Mac>::new_from_slice(&km).expect("hmac key");
    mac.update(canonical_header);
    mac.verify_slice(expected).is_ok()
}

File: go/pigpen/crypto.go
package pigpen

import (
	"crypto/ecdh"
	"crypto/elliptic"
	"crypto/hkdf"
	"crypto/hmac"
	"crypto/rand"
	"crypto/sha256"
	"encoding/binary"
	"errors"
	"fmt"
	"io"

	"golang.org/x/crypto/chacha20poly1305"
)

// pigpen-v1 crypto suite (RFC 0008 §4).

const (
	fileKeyLen      = 16 // RFC 0008 §4.1
	payloadNonceLen = 16 // RFC 0008 §4.5
	streamChunkSize = 64 * 1024
	streamTagLen    = chacha20poly1305.Overhead // 16

	infoPayload = "pigpen-v1 payload"
	infoHeader  = "pigpen-v1 header"
	infoP256    = "pigpen-v1 piv-p256"
	infoX25519  = "pigpen-v1 x25519"
)

// concat returns a fresh slice holding a followed by b, allocated once. Used
// to build the epk‖recipient HKDF salt without aliasing either input.
func concat(a, b []byte) []byte {
	out := make([]byte, 0, len(a)+len(b))
	out = append(out, a...)
	return append(out, b...)
}

func hkdf32(secret, salt []byte, info string) []byte {
	out, err := hkdf.Key(sha256.New, secret, salt, info, 32)
	if err != nil {
		// HKDF over 32 bytes of SHA-256 output never fails for these inputs.
		panic(err)
	}
	return out
}

// aeadSeal/aeadOpen wrap ChaCha20-Poly1305 with a 12-byte nonce.
func aeadSeal(key, nonce, plaintext []byte) ([]byte, error) {
	aead, err := chacha20poly1305.New(key)
	if err != nil {
		return nil, err
	}
	return aead.Seal(nil, nonce, plaintext, nil), nil
}

func aeadOpen(key, nonce, ciphertext []byte) ([]byte, error) {
	aead, err := chacha20poly1305.New(key)
	if err != nil {
		return nil, err
	}
	return aead.Open(nil, nonce, ciphertext, nil)
}

var zeroNonce = make([]byte, chacha20poly1305.NonceSize)

// --- X25519 wrap (RFC 0008 §4.4) -----------------------------------------

func wrapX25519(fileKey, recipientPub []byte, rng io.Reader) (blob []byte, err error) {
	curve := ecdh.X25519()
	esk, err := curve.GenerateKey(rng)
	if err != nil {
		return nil, err
	}
	rpk, err := curve.NewPublicKey(recipientPub)
	if err != nil {
		return nil, err
	}
	shared, err := esk.ECDH(rpk)
	if err != nil {
		return nil, err
	}
	epk := esk.PublicKey().Bytes() // 32 bytes
	salt := concat(epk, recipientPub)
	kw := hkdf32(shared, salt, infoX25519)
	ct, err := aeadSeal(kw, zeroNonce, fileKey)
	if err != nil {
		return nil, err
	}
	return append(epk, ct...), nil // 32 + 32
}

func unwrapX25519(blob, recipientPub, recipientSec []byte) (fileKey []byte, err error) {
	if len(blob) != 32+fileKeyLen+streamTagLen {
		return nil, fmt.Errorf("pigpen: bad x25519 wrap length %d", len(blob))
	}
	epk, ct := blob[:32], blob[32:]
	curve := ecdh.X25519()
	sk, err := curve.NewPrivateKey(recipientSec)
	if err != nil {
		return nil, err
	}
	epub, err := curve.NewPublicKey(epk)
	if err != nil {
		return nil, err
	}
	shared, err := sk.ECDH(epub)
	if err != nil {
		return nil, err
	}
	salt := concat(epk, recipientPub)
	kw := hkdf32(shared, salt, infoX25519)
	return aeadOpen(kw, zeroNonce, ct)
}

// --- P-256 wrap (RFC 0008 §4.3) ------------------------------------------
//
// Encrypt-side is pure software. Decrypt-side delegates the ECDH to an
// ECDHOracle (a card via piggy-agent), so the slot-9D scalar never
// materialises and the WASM module does no PCSC I/O.

func wrapP256(fileKey, recipientCompressed []byte, rng io.Reader) (blob []byte, err error) {
	curve := ecdh.P256()
	esk, err := curve.GenerateKey(rng)
	if err != nil {
		return nil, err
	}
	rpk, err := p256PublicFromCompressed(recipientCompressed)
	if err != nil {
		return nil, err
	}
	shared, err := esk.ECDH(rpk) // 32-byte X-coordinate
	if err != nil {
		return nil, err
	}
	epkCompressed, err := p256Compress(esk.PublicKey().Bytes())
	if err != nil {
		return nil, err
	}
	salt := concat(epkCompressed, recipientCompressed)
	kw := hkdf32(shared, salt, infoP256)
	ct, err := aeadSeal(kw, zeroNonce, fileKey)
	if err != nil {
		return nil, err
	}
	return append(epkCompressed, ct...), nil // 33 + 32
}

func unwrapP256(blob, recipientCompressed []byte, oracle ECDHOracle, self markIdentity) (fileKey []byte, err error) {
	if len(blob) != 33+fileKeyLen+streamTagLen {
		return nil, fmt.Errorf("pigpen: bad p256 wrap length %d", len(blob))
	}
	epkCompressed, ct := blob[:33], blob[33:]
	shared, err := oracle.ECDH(self, epkCompressed) // 32-byte X-coordinate from the card
	if err != nil {
		return nil, err
	}
	salt := concat(epkCompressed, recipientCompressed)
	kw := hkdf32(shared, salt, infoP256)
	return aeadOpen(kw, zeroNonce, ct)
}

// p256 point (de)compression via the stdlib elliptic helpers. RustCrypto's
// p256 crate does this natively; here we lean on elliptic.* which builds
// fine for wasm (deprecation warnings are not errors).

func p256Compress(uncompressed []byte) ([]byte, error) {
	x, y := elliptic.Unmarshal(elliptic.P256(), uncompressed)
	if x == nil {
		return nil, errors.New("pigpen: invalid uncompressed P-256 point")
	}
	return elliptic.MarshalCompressed(elliptic.P256(), x, y), nil
}

func p256PublicFromCompressed(compressed []byte) (*ecdh.PublicKey, error) {
	x, y := elliptic.UnmarshalCompressed(elliptic.P256(), compressed)
	if x == nil {
		return nil, errors.New("pigpen: invalid compressed P-256 point")
	}
	uncompressed := elliptic.Marshal(elliptic.P256(), x, y)
	return ecdh.P256().NewPublicKey(uncompressed)
}

// --- Payload STREAM (RFC 0008 §4.5) --------------------------------------

func sealPayload(fileKey, plaintext []byte, rng io.Reader) ([]byte, error) {
	nonce := make([]byte, payloadNonceLen)
	if _, err := io.ReadFull(rng, nonce); err != nil {
		return nil, err
	}
	streamKey := hkdf32(fileKey, nonce, infoPayload)
	aead, err := chacha20poly1305.New(streamKey)
	if err != nil {
		return nil, err
	}
	out := append([]byte{}, nonce...)
	// Chunk the plaintext; emit at least one (final) chunk so empty
	// plaintext still produces a terminating tag (age parity).
	for i := 0; ; i++ {
		start := i * streamChunkSize
		end := start + streamChunkSize
		last := end >= len(plaintext)
		if end > len(plaintext) {
			end = len(plaintext)
		}
		chunk := plaintext[start:end]
		out = aead.Seal(out, streamNonce(uint64(i), last), chunk, nil)
		if last {
			break
		}
	}
	return out, nil
}

func openPayload(fileKey, payload []byte) ([]byte, error) {
	if len(payload) < payloadNonceLen {
		return nil, errors.New("pigpen: payload shorter than nonce")
	}
	nonce, body := payload[:payloadNonceLen], payload[payloadNonceLen:]
	streamKey := hkdf32(fileKey, nonce, infoPayload)
	aead, err := chacha20poly1305.New(streamKey)
	if err != nil {
		return nil, err
	}
	encChunk := streamChunkSize + streamTagLen
	var out []byte
	for i := 0; ; i++ {
		start := i * encChunk
		if start >= len(body) {
			// We consumed everything but never saw a "last" chunk.
			return nil, errors.New("pigpen: truncated payload (no final chunk)")
		}
		end := start + encChunk
		last := end >= len(body)
		if end > len(body) {
			end = len(body)
		}
		chunk := body[start:end]
		plain, err := aead.Open(nil, streamNonce(uint64(i), last), chunk, nil)
		if err != nil {
			// Retry as a non-final chunk only matters when our "last"
			// guess was wrong; with fixed-size chunking the guess is
			// exact except at the boundary, so a failure here is real.
			return nil, fmt.Errorf("pigpen: chunk %d auth failed: %w", i, err)
		}
		out = append(out, plain...)
		if last {
			break
		}
	}
	return out, nil
}

// streamNonce builds the 12-byte age STREAM nonce: 11-byte big-endian
// counter followed by a 1-byte last-chunk flag.
func streamNonce(counter uint64, last bool) []byte {
	n := make([]byte, 12)
	binary.BigEndian.PutUint64(n[3:11], counter) // top 3 bytes stay zero (88-bit counter)
	if last {
		n[11] = 0x01
	}
	return n
}

// --- Header MAC (RFC 0008 §4.6) ------------------------------------------

func headerMAC(fileKey, canonicalHeader []byte) []byte {
	km := hkdf32(fileKey, nil, infoHeader)
	m := hmac.New(sha256.New, km)
	m.Write(canonicalHeader)
	return m.Sum(nil)
}

func randomFileKey(rng io.Reader) ([]byte, error) {
	fk := make([]byte, fileKeyLen)
	_, err := io.ReadFull(rng, fk)
	return fk, err
}

// defaultRand is the package CSPRNG; tests may inject a deterministic one.
var defaultRand io.Reader = rand.Reader
